### PR TITLE
Tensorboard to get gradients only for trainable weights - updated

### PR DESCRIPTION
**Pull request by SnehaNB**
### Summary
#11334 updated with test case

### PR Overview

- [y] This PR requires new unit tests [y/n] (make sure tests are included)
- [n] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [y] This PR is backwards compatible [y/n]
- [n] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

**Comment by SnehaNB** [comment]
@fchollet not sure what's happening with tests/test_dynamic_trainability.py

File: tests/keras/test_callbacks.py [hunk 1]
```diff
@@ -936,5 +938,47 @@ def tests_RemoteMonitorWithJsonPayload():
...
+    input_shape = (72, 72, 3)
+    (x_train, y_train), (x_test, y_test) = get_test_data(num_train=train_samples,
+                                                         num_test=test_samples,
+                                                         input_shape=input_shape,
+                                                         classification=True,
+                                                         num_classes=num_classes)
+    y_train = np_utils.to_categorical(y_train)
+    y_test = np_utils.to_categorical(y_test)
+
+    base_model = Xception(include_top=False,
```
Comment (fchollet) [inline tests/keras/test_callbacks.py]: This test would be very expensive to run. It's an integration test, not a unit test. Please boil it down to essential components.

**Comment by SnehaNB** [comment]
@fchollet PTAL

File: keras/callbacks.py [hunk 2]
```diff
@@ -790,7 +790,7 @@ def set_model(self, model):
             self.sess = K.get_session()
         if self.histogram_freq and self.merged is None:
             for layer in self.model.layers:
-                for weight in layer.weights:
+                for weight in layer.trainable_weights:
```
Comment (fchollet) [inline keras/callbacks.py]: This is too restrictive. Rather, the conditional below should be `if self.write_grads and weight in layer.trainable_weights`

File: tests/keras/test_callbacks.py [hunk 3]
```diff
@@ -936,5 +938,44 @@ def tests_RemoteMonitorWithJsonPayload():
                   validation_data=(X_test, y_test), callbacks=cbks, epochs=1)
 
 
+def test_TensorBoard_trainable_weights(tmpdir):
+    np.random.seed(np.random.randint(1, 1e7))
+    filepath = str(tmpdir / 'logs')
```
Comment (fchollet) [inline tests/keras/test_callbacks.py]: `os.path.join(tmpdir, 'logs')`

File: tests/keras/test_callbacks.py [hunk 4]
```diff
@@ -936,5 +938,44 @@ def tests_RemoteMonitorWithJsonPayload():
                   validation_data=(X_test, y_test), callbacks=cbks, epochs=1)
 
 
+def test_TensorBoard_trainable_weights(tmpdir):
+    np.random.seed(np.random.randint(1, 1e7))
```
Comment (fchollet) [inline tests/keras/test_callbacks.py]: No need for a seed in this test

File: tests/keras/test_callbacks.py [hunk 5]
```diff
@@ -936,5 +938,44 @@ def tests_RemoteMonitorWithJsonPayload():
...
+    filepath = str(tmpdir / 'logs')
+
+    input_shape = (72, 72, 3)
+    (x_train, y_train), (x_test, y_test) = get_test_data(num_train=train_samples,
+                                                         num_test=test_samples,
+                                                         input_shape=input_shape,
+                                                         classification=True,
+                                                         num_classes=num_classes)
+    y_train = np_utils.to_categorical(y_train)
+    y_test = np_utils.to_categorical(y_test)
```
Comment (fchollet) [inline tests/keras/test_callbacks.py]: In general this test is more like an integration test than a unit test. You don't need half of this stuff:

- the model should be minimal (this one has a bunch of extra layers)
- you don't need data with a statistical structures, `np.random` will work just fine
- etc.

**Comment by SnehaNB** [comment]
Ive made changes to an already existing test. This should suffice hopefully?

**Comment by SnehaNB** [comment]
@fchollet modified an existing test instead

**Comment by fchollet** [comment]
You could also apply the same fix (and test) to `tf.keras` in the TensorFlow repository, that would be very helpful.
